### PR TITLE
feat: poc octo run list

### PR DESCRIPTION
**Pull request by GustavEikaas**
>[!IMPORTANT]
>This is a draft PR
>Im making this draft PR so I can get reviews/suggestions during development
>I will fill out the PR template when the PR is ready for review

Resolves: #149 

### Describe what this PR does / why we need it
First PR in a series of PRs to add support for github actions

### Checklist

- [ ] Passing tests and linting standards
- [ ] Documentation updates in README.md and doc/octo.txt

## Feature checklist
- [x] Buffer
  - [x] Expand/collapse node
  - [x] Highlights
  - [x] Keybindings
    - [x] Refresh
- [x] Workflows
  - [x] Support multiple jobs
- [x] Workflow job steps
  - [x] Logs
  - [x] Status
  - [x] `##[group]`
  - [x] `[command]`

## Remaining

- [x] Customizable keymappings
- [x] Customizable icons
- [ ] Update readme
- [ ] Cleanup code
- [ ] Customizable refresh interval
- [ ] ~Look into using graphql in favor of gh cli commands (not sure if strictly necessary but conforms to the repo standard)~
- [x] Use telescope pickers from the repo
- [ ] Understand the writers.lua file and extend it
- [ ] Use existing octo window creator

### CODE REVIEW

**Comment by wd60622** [comment]
Using graphql is not a big deal. Just use whatever endpoint that works. Think there are a few non graphql commands used.

**Comment by GustavEikaas** [comment]
Ah okay, functionality wise, do you have any feedback?

**Comment by wd60622** [comment]
> Ah okay, functionality wise, do you have any feedback?

I will give it a try when I can!

**Comment by wd60622** [comment]
Think it looks good. Would you want to make use of the pickers? I have been using telescope recently. What do you use?

Some items: 

- What do you expect the callback to be upon selection an item? We can work something in like #642 in order to support more in the future.
- Would there be a preview or buffer to show each item? What would be populated there? Maybe a buffer doesn't make sense here. What are your thoughts?

File: lua/octo/workflow_runs.lua [hunk 1]
```diff
@@ -0,0 +1,533 @@
...
+    month = month,
+    day = day,
+    hour = hour,
+    min = min,
+    sec = sec,
+    isdst = false,
+  }
+end
+
+local function time_ago(timestamp)
```
Comment (wd60622) [inline lua/octo/workflow_runs.lua:230]: There is a similar function in utils which is used in all of the events. Maybe that can be used instead.

**Comment by GustavEikaas** [comment]
image
@wd60622 Is this what you meant, kinda like the PR and issue flow?

File: lua/octo/workflow_runs.lua (hunk 1 above)
Comment (GustavEikaas) [reply lua/octo/workflow_runs.lua]: Ah nice, ill use that instead

**Comment by GustavEikaas** [comment]
You still want there to be a buffer when you press `<CR>` on one of the items in the picker. AKA same flow as pull request?

**Comment by pwntester** [comment]
Thanks for the PR @GustavEikaas!

I like the idea of showing the status of the run in the preview buffer and perhaps show the logs of the run upon opening the item with `<CR>`. The problem I see is that a run may contain multiple jobs and each job will have its own log. Should we present them sequentially? thoughts?

**Comment by GustavEikaas** [comment]
IIRC the sequential jobs is already supported but yes I was thinking the same. 
For the logs of every step we should probably lazy load and use fold or something similiar. Some steps have very long log outputs

**Comment by GustavEikaas** [comment]
@pwntester 
Here is an image of what it looks like with multiple jobs
image

**Comment by GustavEikaas** [comment]
Doesnt seem like there is a good structured output for workflow logs. Will probably be tricky to have the same UI as github web.
Only way I see right now is query the workflow logs and using regex to match the lines to the corresponding steps

**Comment by pwntester** [comment]
Nice, perhaps we can do a nested telescope selection where the use first chooses the run and then the job, that should get us a shorter log to show to the user and it should easier to print since we dont need to care about job's dependencies, just sequential steps

**Comment by GustavEikaas** [comment]
I have been super busy at work so I have not had that much time to look at this but I finally got some time and this is how far I have gotten.

image

## Features
- Telescope picker
- Telescope preview buffer
- Buffer
- `>` indicates a collapsable line
- Multi-job support
- Fetch logs on demand
- Yellow highlight means its a command (the line includes a `[command]`
- `>` inside the log text(purple highlight) means its a group (line include `##[group] and ##[endgroup]`
- Expanding a step fetches the logs for the workflow and assigns corresponding log entries to all jobs and steps


## Considerations
- It's pretty static at the moment and the cache is either 100% static, assign once and read forever. Or no cache, fetch every time invoked. There are a lot of ifs and buts for making a good cache control
- Interactivity. It would be nice for pending jobs to auto-refresh

**Comment by GustavEikaas** [comment]
This PR is being worked on actively, im just testing it. The best way to iron out all bugs is to test it at work where I have big projects with complex github actions. I tried testing with some dummy actions but they dont accurately capture the complexity of real ones. Plus its a huge waste of resources

**Comment by wd60622** [comment]
> This PR is being worked on actively, im just testing it. The best way to iron out all bugs is to test it at work where I have big projects with complex github actions. I tried testing with some dummy actions but they dont accurately capture the complexity of real ones. Plus its a huge waste of resources

Cool. I can check it out and give it a try

**Comment by GustavEikaas** [comment]
That would be great!

This pr grew in scope as I was working on it so im not sure what the best strategy is. I can try and break it into smaller prs ensuring its in a usable state

**Comment by wd60622** [comment]
When using the `o` mapping on run file, I am getting trackback

```
        [C]: in function 'readfile'
        ...tHub/neovim-plugins/octo.nvim/lua/octo/workflow_runs.lua:322: in function 'cb'
        ...tHub/neovim-plugins/octo.nvim/lua/octo/workflow_runs.lua:208: in function 'traverse'
        ...tHub/neovim-plugins/octo.nvim/lua/octo/workflow_runs.lua:210: in function 'traverse'
        ...tHub/neovim-plugins/octo.nvim/lua/octo/workflow_runs.lua:210: in function 'traverse'
        ...tHub/neovim-plugins/octo.nvim/lua/octo/workflow_runs.lua:314: in function 'get_logs'
        ...tHub/neovim-plugins/octo.nvim/lua/octo/workflow_runs.lua:366: in function 'cb'
        ...tHub/neovim-plugins/octo.nvim/lua/octo/workflow_runs.lua:591: in function <...tHub/neovim-plugins/octo.nvim/lua/octo/workflow_runs.lua:
587>
```

File: lua/octo/workflow_runs.lua [hunk 2]
```diff
@@ -0,0 +1,730 @@
...
+  vim.notify "Fetching workflow runs (this may take a while) ..."
+  local co = coroutine.running()
+  local wf_runs = get_workflow_runs_sync(co)
+
+  preview_picker(
+    nil,
+    wf_runs,
+    function(selected)
+      render(selected)
+    end,
```
Comment (wd60622) [inline lua/octo/workflow_runs.lua:713]: Nit. just pass `render`
```suggestion
    render,
```

File: lua/octo/workflow_runs.lua
Comment (wd60622) [inline lua/octo/workflow_runs.lua:1]: Can we do all requires at top. If there is a concern of picker plugins required, then associated logic to `pickers/telescope` folder

File: lua/octo/workflow_runs.lua [hunk 3]
```diff
@@ -0,0 +1,730 @@
...
+      vim.api.nvim_buf_add_highlight(M.buf, namespace, vl.highlight, vl.index, 0, -1)
+    end
+  end
+
+  for binding, cb in pairs(tree_keymaps) do
+    vim.keymap.set("n", binding, function()
+      local current_line = vim.api.nvim_win_get_cursor(0)[1]
+      local line = lines[current_line]
+      if line.node_ref ~= nil then
+        cb(line.node_ref)
```
Comment (wd60622) [inline lua/octo/workflow_runs.lua:591]: Noticing that this could take long (Maybe my internet is bad). Could there be a `vim.notify` notification to tell the user

File: lua/octo/workflow_runs.lua [hunk 4]
```diff
@@ -0,0 +1,730 @@
...
+
+
+-- Accepts zip contents and writes and then unzips them
+---@param stdout string - The zip content to write
+local function write_and_unzip_file(stdout)
+  local temp_location = os.tmpname()
+  local zip_location = temp_location .. ".zip"
+  local file = io.open(zip_location, "wb")
+  if not file then
+    print("Failed to create temporary file")
```
Comment (wd60622) [inline lua/octo/workflow_runs.lua:271]: We can use require("octo.utils").error and info instead of the print statements

**Comment by wd60622** [comment]
Do you have pre-commit install for linting? `pre-commit install`

**Comment by wd60622** [comment]
I don't have icons next to each step. Maybe the buffer is not completely rendering for me

File: lua/octo/workflow_runs.lua
Comment (GustavEikaas) [reply lua/octo/workflow_runs.lua:1]: Done

File: lua/octo/workflow_runs.lua (hunk 3 above)
Comment (GustavEikaas) [reply lua/octo/workflow_runs.lua:587]: Are you referring to the get workflow logs part?

**Comment by GustavEikaas** [comment]
> I don't have icons next to each step. Maybe the buffer is not completely rendering for me

I went a bit back and forth on this because there is so much visual noise. I think success icons doesnt make sense as that is the expected outcome? Would love to hear some opinions on this

image
Here is an example of one where the first job fails and then the second skips because the first one failed

image
Here is an example of one where every step passes

**Comment by GustavEikaas** [comment]
I dont know how common nerdfonts are but if users have nerdfonts installed the expanded, collapsed chevrons look way better
image

File: lua/octo/workflow_runs.lua (hunk 3 above)
Comment (GustavEikaas) [reply lua/octo/workflow_runs.lua:597]: Added a loading message for retrieving workflow logs

**Comment by GustavEikaas** [comment]
> When using the `o` mapping on run file, I am getting trackback
> 
> ```
> Failed to unzip logs: checkdir:  cannot create extraction directory: /tmp/lua_3VxzuD
>            File exists
> E5108: Error executing lua: Vim:E484: Can't open file /tmp/lua_3VxzuD/update-comment/1_Set up job.txt
> stack traceback:
>         [C]: in function 'readfile'
>         ...tHub/neovim-plugins/octo.nvim/lua/octo/workflow_runs.lua:322: in function 'cb'
>         ...tHub/neovim-plugins/octo.nvim/lua/octo/workflow_runs.lua:208: in function 'traverse'
>         ...tHub/neovim-plugins/octo.nvim/lua/octo/workflow_runs.lua:210: in function 'traverse'
>         ...tHub/neovim-plugins/octo.nvim/lua/octo/workflow_runs.lua:210: in function 'traverse'
>         ...tHub/neovim-plugins/octo.nvim/lua/octo/workflow_runs.lua:314: in function 'get_logs'
>         ...tHub/neovim-plugins/octo.nvim/lua/octo/workflow_runs.lua:366: in function 'cb'
>         ...tHub/neovim-plugins/octo.nvim/lua/octo/workflow_runs.lua:591: in function <...tHub/neovim-plugins/octo.nvim/lua/octo/workflow_runs.lua:
> 587>
> ```

Fixed in latest commit by deleting the files after reading them

**Comment by GustavEikaas** [comment]
Okay so i'm kinda stuck with this PR. I can easily finish it but it seems impossible to get the optimal experience.

One of the most essential parts of github actions is for me the logs. I usually do one of three things. 
1. Follow a running job, (track the stdout of the current step in near realtime)
2. Investigate a failed job
3. Check out a workflow run that has completed. Possibly looking at the stdout

Doing 2 and 3 is no problem but doing 1 has several blockers. The zip archive im requesting from github api only works for completed jobs. AFAIK there is no way to get structured output for a running job. If I were to query logs for a running job there will be 2 main obstacles.

1. Rate limiting. There is no way to subscribe to logs coming from a runner so this means spamming and endpoint requesting the raw logs for an entire workflow. You shouldnt and probably cannot do this very often. Maybe 10sec intervals could work(?)

2. Matching job/steps to correct output. Same issue as previously mentioned in this thread. Without the zip archive we are back to lua patterns and some fine guessing. Also includes the duplicate step names within same job problems.


We can either skip implementing 1 for now, which would be merging something usable but with a major missing feature. Or I can do a "double implementation". Where i use running logs for running jobs and zip archive when a job completes

(also sorry this has taken so long, been really busy at work. Its also hard to keep the spark going when you run into major roadblocks)

@wd60622 @pwntester Thoughts?

**Comment by wd60622** [comment]
No worries. The experience doesn't have to be optimal. I just a working solution for *finished* logics would be fine. No need to aim for streaming. As you mentioned, looking at failed jobs is likely the most common usecase.

Let's try to get this merged in and people can try. 

For me, it still doesn't allow me to open. I am getting that same traceback.

What is your current bandwidth for this PR?

**Comment by GustavEikaas** [comment]
If the goal is to prioritize looking at failed jobs I can probably finish the PR this week. Ill look closer into your issue after work. Looks like os.tmpname fails to generate a unique name for you. weird...

**Comment by GustavEikaas** [comment]
@wd60622 After looking closer at the code im still thinking it fails to generate a unique tmpname which is pretty weird. I'm also using WSL but no issue for me.. Anyways I updated the code to handle the case of generating a unique tmpname. Would you mind testing again. Other than that I think the code is pretty much done, just a bunch of CR fixes remaining presumably. Feel free to review and comment the code

**Comment by wd60622** [comment]
I've just updated my branch and still not working :( I will have to investigate further. Same traceback as before.

**Comment by GustavEikaas** [comment]
Hmm if youre getting exactly the same traceback as before I think it means the plugin isnt updated with the latest commit. Could you try removing it and installing it again from the same branch? Im pretty sure I have ran into the same issue with lazy before, Update doesnt always work

**Comment by wd60622** [comment]
This is my configuration. It should be using the latest version of your branch : 

https://github.com/wd60622/dotfiles/blob/9b9dfdb2268a34f9c699c46fe89659ce6bdebfc0/nvim/lua/plugins/version-control.lua#L204-L204

**Comment by GustavEikaas** [comment]
Does it reference the same tmpname in the stacktrace every time?

**Comment by wd60622** [comment]
I was mistaken. This is a different traceback: 

```
E5108: Error executing lua: ...tHub/neovim-plugins/octo.nvim/lua/octo/workflow_runs.lua:281: Failed to create a unique temporary file name after 5
0 attempts.
stack traceback:
        [C]: in function 'error'
        ...tHub/neovim-plugins/octo.nvim/lua/octo/workflow_runs.lua:281: in function 'create_unique_tmpname'
        ...tHub/neovim-plugins/octo.nvim/lua/octo/workflow_runs.lua:287: in function 'write_and_unzip_file'
        ...tHub/neovim-plugins/octo.nvim/lua/octo/workflow_runs.lua:342: in function 'get_logs'
        ...tHub/neovim-plugins/octo.nvim/lua/octo/workflow_runs.lua:402: in function 'cb'
        ...tHub/neovim-plugins/octo.nvim/lua/octo/workflow_runs.lua:623: in function <...tHub/neovim-plugins/octo.nvim/lua/octo/workflow_runs.lua:
619>
E5108: Error executing lua: ...tHub/neovim-plugins/octo.nvim/lua/octo/workflow_runs.lua:281: Failed to create a unique temporary file name after 5
0 attempts.
stack traceback:
        [C]: in function 'error'
        ...tHub/neovim-plugins/octo.nvim/lua/octo/workflow_runs.lua:281: in function 'create_unique_tmpname'
        ...tHub/neovim-plugins/octo.nvim/lua/octo/workflow_runs.lua:287: in function 'write_and_unzip_file'
        ...tHub/neovim-plugins/octo.nvim/lua/octo/workflow_runs.lua:342: in function 'get_logs'
        ...tHub/neovim-plugins/octo.nvim/lua/octo/workflow_runs.lua:402: in function 'cb'
        ...tHub/neovim-plugins/octo.nvim/lua/octo/workflow_runs.lua:623: in function <...tHub/neovim-plugins/octo.nvim/lua/octo/workflow_runs.lua:
619>
E5108: Error executing lua: ...tHub/neovim-plugins/octo.nvim/lua/octo/workflow_runs.lua:281: Failed to create a unique temporary file name after 5
0 attempts.
stack traceback:
        [C]: in function 'error'
        ...tHub/neovim-plugins/octo.nvim/lua/octo/workflow_runs.lua:281: in function 'create_unique_tmpname'
        ...tHub/neovim-plugins/octo.nvim/lua/octo/workflow_runs.lua:287: in function 'write_and_unzip_file'
        ...tHub/neovim-plugins/octo.nvim/lua/octo/workflow_runs.lua:342: in function 'get_logs'
        ...tHub/neovim-plugins/octo.nvim/lua/octo/workflow_runs.lua:402: in function 'cb'
        ...tHub/neovim-plugins/octo.nvim/lua/octo/workflow_runs.lua:623: in function <...tHub/neovim-plugins/octo.nvim/lua/octo/workflow_runs.lua:
619>

```

**Comment by GustavEikaas** [comment]
@wd60622 That is crazy, here is the code. Looks a lot to me like your tmp folder is filled with an insane amount of files or there is something wrong with os.tmpname implementation?

```lua
local function create_unique_tmpname()
  local depth = 50
  local attempts = 0
  local tmpname

  while attempts < depth do
    tmpname = os.tmpname()
    local file = io.open(tmpname, "r")

    if not file then
      return tmpname
    else
      file:close()
    end

    attempts = attempts + 1
  end

  error("Failed to create a unique temporary file name after " .. depth .. " attempts.")
end

```

**Comment by wd60622** [comment]
Yeah, weird. Thanks for sharing the snippet. Let me isolate it and see what happens

**Comment by wd60622** [comment]
The files are being created in the /tmp folder. I see them after running the function isolated.

**Comment by GustavEikaas** [comment]
hmm problem with your edited snippet is im preparing a new directory for unzip to unzip into. If I write to that file it will exist and thus cannot be overwritten. I need the function to generate a unique name for a file that does not yet exist. Whats weird however is how reading or writing the file makes the problem go away for you. Strange...

**Comment by wd60622** [comment]
It seems to create the file on read. Is there an directory exists function that could be used instead? Maybe in plentary

**Comment by GustavEikaas** [comment]
If i put this in my init.lua function in throws an error when i start neovim. This is because the file does not exist and thus the filehandle is nil. But the following code seems to behave different for you and create the file?
```lua
local tmpname = os.tmpname()
local file = io.open(tmpname, "r")
file:close();
print(tmpname)
```

I guess i could replace the code with, expecting success == false
local tmpname = os.tmpname()
local success = pcall(vim.fn.readfile, tmpname)

**Comment by GustavEikaas** [comment]
A lot of testing for you but maybe the latest one works? where i use vim.fn.readfile instead to check if file exists?

**Comment by GustavEikaas** [comment]
Finally found it. Its a POSIX quirk. From the lua docs for os.tmpname
```md
In POSIX systems, this function also creates a file with that name, to avoid security risks. (Someone else might create the file with wrong permissions in the time between getting the name and creating the file.) You still have to open the file to use it and to remove it (even if you do not use it).
```

**Comment by GustavEikaas** [comment]
Latest commit works fine on arch. Ill test on windows aswell just to make sure

**Comment by wd60622** [comment]
Thanks for the updates. will give it a try when I can

**Comment by wd60622** [comment]
Okay. I am able to get it to work (for a few) now. I am still getting this traceback: 

```
E5108: Error executing lua: Vim:E484: Can't open file vpavbvq/X64-ubuntu (ubuntu-20.04, https:/github.com/neovim/neovim/releases/download/v0.7.0/n
vim-linux64..../1_Set up job.txt
stack traceback:
        [C]: in function 'readfile'
        ...tHub/neovim-plugins/octo.nvim/lua/octo/workflow_runs.lua:341: in function 'cb'
        ...tHub/neovim-plugins/octo.nvim/lua/octo/workflow_runs.lua:208: in function 'traverse'
        ...tHub/neovim-plugins/octo.nvim/lua/octo/workflow_runs.lua:210: in function 'traverse'
        ...tHub/neovim-plugins/octo.nvim/lua/octo/workflow_runs.lua:210: in function 'traverse'
        ...tHub/neovim-plugins/octo.nvim/lua/octo/workflow_runs.lua:333: in function 'get_logs'
        ...tHub/neovim-plugins/octo.nvim/lua/octo/workflow_runs.lua:390: in function 'cb'
        ...tHub/neovim-plugins/octo.nvim/lua/octo/workflow_runs.lua:611: in function <...tHub/neovim-plugins/octo.nvim/lua/octo/workflow_runs.lua:
607>
```

**Comment by GustavEikaas** [comment]
Yeah ill take a last look at escaping/normalizing filenames

**Comment by wd60622** [comment]
> Yeah ill take a last look at escaping/normalizing filenames

Great. Thank you!

**Comment by wd60622** [comment]
I am also noticing that the folders are being created off of my cwd. Is that as expected?

**Comment by GustavEikaas** [comment]
Good catch ill fix that, happened after i migrated away from os.tmpname.

**Comment by GustavEikaas** [comment]
Encountered a strange scenario I have never seen before. In a workflow from Neovim repo it failed to load because workflow.jobs == nil. After opening it in browser this is what I see. 
image
Given the data available to me this is the best I can do. Clearly the gh api is not as powerful as the web UI. But it will hopefully improve over time.
image

Im not sure if I fixed the same issue you were facing but for me it would fail if it tried to read the logs of a step that was skipped. This was because the .txt file is missing from the zip archive. I added a check so that skipped steps are not assigned logs.

**Comment by GustavEikaas** [comment]
Still a couple of rough edges in this PR but I think its getting ready to merge. Given that it stops giving you errors

For any future issues/prs/discussions feel free to tag me, I would like to keep working on this

**Comment by wd60622** [comment]
I'm having a lot better luck with it now. Really cool feature. 

I am running into this: 

```
                                                                                                                                1,15          All
E5108: Error executing lua: ...tHub/neovim-plugins/octo.nvim/lua/octo/workflow_runs.lua:597: 'replacement string' item contains newlines
stack traceback:
        [C]: in function 'nvim_buf_set_lines'
        ...tHub/neovim-plugins/octo.nvim/lua/octo/workflow_runs.lua:597: in function 'print_lines'
        ...tHub/neovim-plugins/octo.nvim/lua/octo/workflow_runs.lua:623: in function 'refresh'
        ...tHub/neovim-plugins/octo.nvim/lua/octo/workflow_runs.lua:525: in function 'populate_preview_buffer'
        ...tHub/neovim-plugins/octo.nvim/lua/octo/workflow_runs.lua:680: in function 'define_preview'
        ...scope.nvim/lua/telescope/previewers/buffer_previewer.lua:451: in function 'preview'
        ...share/nvim/lazy/telescope.nvim/lua/telescope/pickers.lua:1212: in function 'refresh_previewer'
        ...share/nvim/lazy/telescope.nvim/lua/telescope/pickers.lua:1153: in function 'set_selection'
        ...share/nvim/lazy/telescope.nvim/lua/telescope/pickers.lua:912: in function 'move_selection'
        ...e/nvim/lazy/telescope.nvim/lua/telescope/actions/set.lua:39: in function 'run_replace_or_original'
        ...re/nvim/lazy/telescope.nvim/lua/telescope/actions/mt.lua:65: in function 'shift_selection'
        .../nvim/lazy/telescope.nvim/lua/telescope/actions/init.lua:87: in function 'run_replace_or_original'
        ...re/nvim/lazy/telescope.nvim/lua/telescope/actions/mt.lua:65: in function 'key_func'
        ...hare/nvim/lazy/telescope.nvim/lua/telescope/mappings.lua:293: in function <...hare/nvim/lazy/telescope.nvim/lua/telescope/mappings.lua:
292>
```

Also, could you bind <C-y> to the buffer to get the URL of the run? Think that could be pretty convenient. Or at least in the picker if it isn't already!

**Comment by GustavEikaas** [comment]
I did bind <C-b> open in browser for the buffer. I can do <C-y> aswell. 

Are you testing against open repositories, can you link a workflow run that fails? Seems like it complains about `\n` in a line being printed. Not really sure why that would happen, weird... Did it happen when printing the workflow or when getting the workflow step logs?

**Comment by GustavEikaas** [comment]
I managed to repro your issue @wd60622 and I might have found a bug in github...
when downloading the archive from [workflow](https://github.com/pwntester/octo.nvim/actions/runs/12967908686)
I get this. But when trying to cd into the directory I get this error
image
image
image


Not really sure how to proceed here. Will make an issue with github to gather some more info. I enabled longpaths on my windows system and still not working

**Comment by wd60622** [comment]
Was still having issues with this. Does it have something to do with the file / folder names. Should we use a hash or something?

**Comment by GustavEikaas** [comment]
Darn, really?

Well the problem is I do api calls to get the workflow and based on the dynamic info there I request an archive with an arbitary number of files with names that can be built up using the workflow but with `x` amount of replaced characters. Ill do some more digging

**Comment by wd60622** [comment]
Thanks for all of the digging. Feel free to ping me to try again.

**Comment by GustavEikaas** [comment]
Ill keep digging! It would be easier if anyone could link me a specific workflow run that causes it to error out

File: lua/octo/workflow_runs.lua [hunk 5]
```diff
@@ -0,0 +1,701 @@
...
+end
+
+local function update_job_details(id)
+  ---@type WorkflowRun
+  local job_details = {}
+  if M.wf_cache[id] ~= nil then
+    M.refresh()
+    return
+  end
+  local cmd = string.format("gh run view %s --json %s", id, fields)
```
Comment (wd60622) [inline lua/octo/workflow_runs.lua:504]: Possible to use existing code here? 

Might look like this:

```lua
gh = require "octo.gh"

gh.run.view {
  id,
  json=fields, 
  opts = {
    cb = function(output, stderr)
    end,
  }
}
```

Comment (GustavEikaas) [reply lua/octo/workflow_runs.lua]: You mean like this?
Or did you want me to create the `gh.run.view`?

```
  gh.run {
    args = { "run", "view", id, "--json", fields },
    cb = function(output, stderr)
      if stderr and not utils.is_blank(stderr) then
        vim.api.nvim_err_writeln(stderr)
        octo_error("Failed to get workflow run for " .. id)
      elseif output then
        job_details = vim.fn.json_decode(output)
        M.wf_cache[id] = job_details
        M.current_wf = job_details
        M.tree = generate_workflow_tree(job_details)
        M.refresh()
      end
    end,
  }


```

File: lua/octo/workflow_runs.lua [hunk 6]
```diff
@@ -0,0 +1,701 @@
...
+  if result.code ~= 0 then
+    error("Failed to execute 'gh repo view': " .. (result.stderr or "Unknown error"))
+  end
+
+  local ok, decoded = pcall(vim.json.decode, result.stdout)
+  if not ok or not decoded or not decoded.nameWithOwner then
+    error "Failed to parse repository name from 'gh repo view' output"
+  end
+
+  return decoded.nameWithOwner
```
[the remark this answers is not in the document]
Comment (GustavEikaas) [reply lua/octo/workflow_runs.lua]: Done

File: lua/octo/workflow_runs.lua (hunk 5 above)
Comment (wd60622) [reply lua/octo/workflow_runs.lua]: Arbitrary CLI commands are already supported with this logic: 

https://github.com/pwntester/octo.nvim/blob/c689dd516a60665958af9e4195d3bc4a9d5c573b/lua/octo/gh/init.lua?plain=1#L329-L367 

So no need to write it. Just use what I wrote and fill in the callback

File: lua/octo/workflow_runs.lua [hunk 7]
```diff
@@ -0,0 +1,686 @@
...
+end
+
+M.previewer = function(self, entry)
+  local id = entry.value.id
+  M.buf = self.state.bufnr
+  populate_preview_buffer(id, self.state.bufnr)
+end
+
+M.list = function()
+  vim.notify "Fetching workflow runs (this may take a while) ..."
```
Comment (wd60622) [inline lua/octo/workflow_runs.lua:672]: Use utils.info over vim.notify

File: lua/octo/workflow_runs.lua [hunk 8]
```diff
@@ -0,0 +1,686 @@
...
+  local co = coroutine.running()
+  local wf_runs = get_workflow_runs_sync(co)
+
+  require("octo.pickers.telescope.provider").workflow_runs(wf_runs, "Workflow runs", render)
```
Comment (wd60622) [inline lua/octo/workflow_runs.lua:676]: Use require "octo.picker" instead. Feel free to have M.not_implemented for the other two pickers

File: lua/octo/workflow_runs.lua [hunk 9]
```diff
@@ -0,0 +1,686 @@
...
+M.refresh = function()
+  print_lines()
+end
+
+local workflow_limit = 100
+
+local function get_workflow_runs_sync(co)
+  local lines = {}
+  vim.fn.jobstart(
+    "gh run list --json conclusion,displayTitle,event,headBranch,name,number,status,updatedAt,databaseId -L "
```
Comment (wd60622) [inline lua/octo/workflow_runs.lua:619]: Possible to use gh.run / gh.run.list here as well? 
Using that ensures that the environment variables are correct

**Review by wd60622 — COMMENTED** [review]
Some comments. 

Also, been thinking about the use of zip files (Still don't have working in all cases for me)...
Do the logs have to stored to file system? There are other items that are just stored in the local variable `M`. Can that strategy also be used?

File: lua/octo/workflow_runs.lua (hunk 7 above)
Comment (GustavEikaas) [reply lua/octo/workflow_runs.lua]: done

File: lua/octo/workflow_runs.lua (hunk 8 above)
Comment (GustavEikaas) [reply lua/octo/workflow_runs.lua]: done

**Comment by GustavEikaas** [comment]
> Also, been thinking about the use of zip files (Still don't have working in all cases for me)...
Do the logs have to stored to file system? There are other items that are just stored in the local variable M. Can that strategy also be used?


Github supports two ways of giving workflow logs. 
1. as one big .txt file, I tried this first but it isnt structured well enough that I can map it correctly to the jobs and steps. 
2. As a zip archive. With corresponding .txt files per step where the structure is `zipRoot/<jobname>/stepname.txt`

I dont necessarily have to write the zip archive to disk but I need some sort of tool/library to extract the files from the zip archive for me. There is no inbuilt lua/vim support for zip archives. I chose unzip because it is included when installing git. AFAIK unzip doesnt accept inline zip archive as stdio. If you know of any other ways to do this im all ears.

I looked a bit more into the issue and I can no longer repro the issue but apparently you can. Do you have any repro on a public OSS repo where you can link me the workflow run. If I can repro the issue I can probably fix it

**Comment by wd60622** [comment]
Overall, it seems to be working fairly well for most cases. I am having trouble with:

**Comment by wd60622** [comment]
Some other nice to have features: 

- Have the `>` become `∨` when expanded for a better visual representation
- Have the `o` mapping close the node from within the line range. For instance, cursor in the logs being shown.

**Comment by wd60622** [comment]
Other than that, think this is pretty much good to go. We can get it out, people can try, and we iterate! It's awesome addition so thanks for all of the back and forth!

**Comment by GustavEikaas** [comment]
I see the issue with the workflow run you linked @wd60622 
`X64-ubuntu (ubuntu-20.04, httpsgithub.comneovimneovimreleasesdownloadv0.10.0nvim-linux64.../3_Run date +%F  todays-date.txt` - API
`X64-ubuntu (ubuntu-20.04, httpsgithub.comneovimneovimreleasesdownloadv0.10.0nvim-linux64../3_Run date +%F  todays-date.txt` - Archive

TLDR; main issue is length being truncated. Honestly dont know how to work around this without making a bunch of code for processing and transforming paths and filenames. I'm afraid it will never be 100% with this approach

**Comment by GustavEikaas** [comment]
Made a final attempt using some wildcard tricks. @wd60622.
Do you want to do a final test and see if you can read workflow logs without any errors now?

**Comment by wd60622** [comment]
Will take a look tomorrow and let you know how it goes

Is there a Json payload that comes back too? Maybe we can use some type of id instead of a raw name

**Review by wd60622 — APPROVED** [review]
Overall, more successes than failures! But let's get this through. Awesome feature! 🚀

**Comment by GustavEikaas** [comment]
Thanks! It's been a journey, mostly fun though. Glad we could finally release it!

Feel free to tag me in any future issues relating to this.

**Comment by wd60622** [comment]
Totally, I know how it goes. Thanks for sticking in there. 

Feel free to open Issues for any features as well.
